internal/leadership/elector.go

```diff
@@ -62,206 +89,294 @@ type Elector struct {
+				e.Logger.Error(e.Name+": Unable to unmarshal leadership notification", "client_id", e.clientID, "err", err)
+				return
+			}
 
-		select {
-		case <-ctx.Done():
-			return
-		case leadershipNotificationChan <- struct{}{}:
-		}
-	}
+			e.Logger.Info(e.Name+": Received notification from notifier", "action", notification.Action, "client_id", e.clientID)
 
-	sub, err := notifier.ListenRetryLoop(ctx, &e.BaseService, e.notifier, notifier.NotificationTopicLeadership, handleNotification)
-	if err != nil { //nolint:staticcheck
-		// TODO(brandur): Propagate this after refactor.
-	}
-	if sub != nil {
-		defer sub.Unlisten(ctx)
-	}
+			if notification.Action != "resigned" || notification.Name != e.instanceName {
+				// We only care about resignations on because we use them to preempt the
+				// election attempt backoff. And we only care about our own key name.
+				return
+			}
 
-	for {
-		if success := e.gainLeadership(ctx, leadershipNotificationChan); !success {
 			select {
 			case <-ctx.Done():
 				return
-			default:
-				// TODO: proper backoff
-				e.logger.Error("gainLeadership returned unexpectedly, waiting to try again")
-				time.Sleep(time.Second)
-				continue
+			case e.leadershipNotificationChan <- struct{}{}:
 			}
 		}
 
-		// notify all subscribers that we're the leader
-		e.notifySubscribers(true)
+		e.Logger.Info(e.Name+": Listening for leadership changes", "client_id", e.clientID, "topic", notifier.NotificationTopicLeadership)
 
-		err := e.keepLeadership(ctx, leadershipNotificationChan)
-		e.notifySubscribers(false)
-		if err != nil {
-			select {
-			case <-ctx.Done():
+		var err error
+		sub, err = notifier.ListenRetryLoop(ctx, &e.BaseService, e.notifier, notifier.NotificationTopicLeadership, handleNotification)
+		if err != nil { //nolint:staticcheck
+			// TODO(brandur): Propagate this after refactor.
+		}
+	}
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		e.Logger.InfoContext(ctx, e.Name+": Run loop started")
+		defer e.Logger.InfoContext(ctx, e.Name+": Run loop stopped")
+
+		if sub != nil {
+			defer sub.Unlisten(ctx)
+		}
+
+		for {
+			if err := e.attemptGainLeadershipLoop(ctx); err != nil {
+				// Function above only returns an error if context was cancelled
+				// or overall context is done.
+				if !errors.Is(err, context.Canceled) && ctx.Err() == nil {
+					panic(err)
+				}
 				return
-			default:
-				// TODO: backoff
-				e.logger.Error("error keeping leadership", "err", err)
-				continue
+			}
+
+			e.Logger.Info(e.Name+": Gained leadership", "client_id", e.clientID)
+			e.testSignals.GainedLeadership.Signal(struct{}{})
+
+			err := e.keepLeadershipLoop(ctx)
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+
+				if !errors.Is(err, errLostLeadership) {
+					e.Logger.Error(e.Name+": Error keeping leadership", "client_id", e.clientID, "err", err)
```

Comment:
Is this inverted? The message seems to align with when the node _lost_ its leadership

Reply:
A little hard to read, but the idea is to log any error that's not a reelection bid loss.

Changed the code so that:

* Inverted the condition and put in a `continue`, with the log statement falling to below the conditional.
* Renamed the error `errLostLeadershipReelection` for better clarity.